### PR TITLE
Revamp the Scoreboard

### PR DESCRIPTION
**Pull request by ghost**
Because Goofball is really nice, but half the shit on his scoreboard doesn't work, and 5000 points every round because the Botanist or Miner did their jobs is starting to get tiring

New Scoreboard items : 
- Litter : Anything pathed under item/weapon/trash (things like empty chips bags, plates, trays, etc) will count towards litter. Litter costs one per item and prevents Ultra-Clean station if on main station (Protip : Mass Driver)
- Silicon Deaths : Counts for two and a half human deaths. You better be sure about that. Counts MoMMIs, does not count AI (they are counted separately)
- Meals Made : Not new, but was commented out. Every time an item is created via the make_food proc (used by microwaves, notably), a point is scored. Might not count some special cooking for now, like making special dishes, deep-frying or such
- Time : Actually spending time on the station will award you points, you get 1 point every five seconds. This adds up to a groundbreaking (not really) 720 points per hour
- Explosions : Every time something explodes, it adds up. Does not affect the score (under Weird section)

Removed Scoreboard items : 
- Research Done : Never worked, and couldn't figure out how to make it work smartly
- Items Shipped : Doesn't work, especially ever since Old Economy got fucked up
- ALL GAMEMODE-RELATED ITEMS (for score, not the general "X was Y") : The way those items are coded is beyond my comprehension. I think it's just better off if I comment them out to prevent round-end lag, you don't need a fancy scoreboard to know that your station just exploded

Modified Scoreboard items : 
- Human deaths, Silicon deaths, AI deaths : Now called on respective mob/death(), ensuring deaths are always caught even if the mob is gibbed. Note that I am aware that this still counts even if the mob is revived/rebuilt/restored, but this is all the more reasons for Medical to get really good. Also now counts deaths everywhere, incidentally, as long as the dead mob had a mind, aka player (easy on the bus, admins)
- Diseases : Attempting something else, due to all sorts of issues I couldn't make sure if it actually register viruses correctly
- Mess : Now goes through a lot more "messes", from egg smudges and ashes to mucus and liquid fuel
- Events Endured : Now works for random events by simply calling when an event is called. Does not work with manual events (semi-intentionally)
- Power Alarms : Now triggers if an APC has a charge below 30 % (down from 95 %), note that APC must be on station and must have a battery (which incidentally excludes the Ghetto Bar if never repaired, bet you didn't even know that)
- Escapees : Now uses a much less retarded proc to check for shuttle escapees and properly counts pods (it didn't at all before)
- Mined Ore : Renamed Smelted Ore, counts ore smelted through ore redemption, but now also the smelter. One per chunk consumed
- Hydroponics Harvests : Now counts one per unit of any plant produced

Misc : 
- Formatting and all, fun stuff
- I did not check gamemode-specific items for all sorts of reasons. Feel free to check them if you work on the corresponding gamemode. Nuke Ops in particular gave me eye cancer
- Scoreboard formatting modified to clearly separate score categories
- Bad items now cost more (especially deaths, diseases and power losses), escapees award more, harvests and mining award far less (1-1 from 5-2. Trust me, you can produce a lot with even the slightest amount of work)
- Bonus items now award a LOT more (Perfect Power is 2500, Ultra-Clean is 5000)

A good example of the new score mechanic : 30 people escaped, 15 people died, the station has a fair amount of mess (150 "units") and litter (20 "units"), three random events happened, 1 hour 30 minutes went by, no diseases, 500 hydro harvests, 200 ore smelted, the AI is online, one borg was destroyed, 50 meals were made

In order : 30 \* 50 - 15 \* 100 - 150 - 20 + 3 \* 200 + 0.2 \* 3600 + 0.2 \* 600 + 0 + 500 + 200 - 250 + 50

The final score would be 1770 (Rating : "Promotions for Everyone"), pretty good round despite the high casualty rate (due to all the people who escaped)

Discussions go below. I intentionally omitted the changelog because the round-end scoreboard is not game-changing

### CODE REVIEW

**Comment by PJB3005** [comment]
Why the hell are messes and litter different

**Comment by dylanstrategie** [comment]
Because they are counted differently, and I felt it would be good for clarity

File: code/game/gamemodes/scoreboard.dm [hunk 1]
```diff
@@ -112,114 +106,173 @@
...
+			continue
+		for(var/obj/item/weapon/cell/C in A.contents)
+			if(C.percent() < 30)
+				score["powerloss"] += 1 //Enough to auto-cut equipment, so alarm
+
+	var/roundlength = world.time/10 //Get a value in seconds
+	score["time"] = round(roundlength) //One point for every five seconds. One minute is 12 points, one hour 720 points
+
+	//Check how many uncleaned mess are on the station. We can't run through cleanable for reasons, so yeah, long
+	for(var/obj/effect/decal/cleanable/M in world)
```
Comment (ComicIronic) [inline code/game/gamemodes/scoreboard.dm]: WORLD

Comment (ComicIronic) [inline code/game/gamemodes/scoreboard.dm]: Also change this to use a messy() proc or something for OOP.

File: code/game/gamemodes/scoreboard.dm [hunk 2]
```diff
@@ -112,114 +106,173 @@
...
+			score["mess"]++
+		if(istype(M, /obj/effect/decal/cleanable/egg_smudge))
+			score["mess"]++
+		if(istype(M, /obj/effect/decal/cleanable/pie_smudge))
+			score["mess"]++
+		if(istype(M, /obj/effect/decal/cleanable/soot))
+			score["mess"]++
+
+
+	for(var/obj/item/trash/T in world)
```
Comment (ComicIronic) [inline code/game/gamemodes/scoreboard.dm]: WORLD

File: code/game/objects/items/trash.dm [hunk 3]
```diff
@@ -89,3 +89,5 @@
 
 /obj/item/trash/attack(mob/M as mob, mob/living/user as mob)
 	return
+
+/obj/item/trash/Destroy()
```
Comment (ComicIronic) [inline code/game/objects/items/trash.dm]: But why

**Comment by ComicIronic** [comment]
Rename ore mined to ore smelted.

**Comment by dylanstrategie** [comment]
This has been done in the latest push

File: code/game/objects/items/trash.dm (hunk 3 above)
Comment (dylanstrategie) [inline code/game/objects/items/trash.dm]: Now serves a purpose for trash_items list

File: code/game/gamemodes/scoreboard.dm [hunk 4]
```diff
@@ -1,10 +1,36 @@
+//Pretty compact file, so we'll do this up here
+/datum/controller/gameticker/proc/messcheck(var/obj/effect/decal/cleanable/M)
+
+	if(istype(M, /obj/effect/decal/cleanable/blood))
```
Comment (clusterfack) [inline code/game/gamemodes/scoreboard.dm]: Do this shit with a proc on /obj/effect/decal/cleanable

File: code/game/gamemodes/scoreboard.dm [hunk 5]
```diff
@@ -33,76 +59,70 @@
...
+				if(findtext(thing, "<font color='orange'>")) //I just dropped 10 IQ points from seeing this
+					score["clownabuse"]++
 
 	for(var/mob/living/player in mob_list)
-		if (player.client)
-			if (player.stat != 2)
-				var/turf/location = get_turf(player.loc)
-				var/area/escape_zone = locate(/area/shuttle/escape/centcom)
-				if (location in escape_zone)
+		if(player.client)
```
Comment (clusterfack) [inline code/game/gamemodes/scoreboard.dm]: Use the player list for this

File: code/game/gamemodes/scoreboard.dm [hunk 6]
```diff
@@ -33,76 +59,70 @@
...
+			if(player.stat != DEAD)
+				var/turf/T = get_turf(player)
+				if(istype(T.loc, /area/shuttle/escape/centcom) || istype(T.loc, /area/shuttle/escape_pod1/centcom) || istype(T.loc, /area/shuttle/escape_pod2/centcom) || istype(T.loc, /area/shuttle/escape_pod3/centcom) || istype(T.loc, /area/shuttle/escape_pod5/centcom))
```
Comment (clusterfack) [inline code/game/gamemodes/scoreboard.dm]: MFW you use istype(specialsnowflake) for EVERYTHING

File: code/game/gamemodes/scoreboard.dm [hunk 7]
```diff
@@ -33,76 +59,70 @@
...
+	//Score Calculation and Display
 
-	// Who is alive/dead, who escaped
-	for (var/mob/living/silicon/ai/I in mob_list)
-		if (I.stat == 2 && I.z == 1)
-			score["deadaipenalty"] = 1
-			score["deadcrew"] += 1
+	//Run through humans for diseases, also the Clown
+	for(var/mob/living/carbon/human/I in mob_list)
+		if(I.viruses) //Do this guy have any viruses ?
```
Comment (clusterfack) [inline code/game/gamemodes/scoreboard.dm]: && viruses.len

File: code/game/gamemodes/scoreboard.dm [hunk 8]
```diff
@@ -33,76 +59,70 @@
...
 			foecount++
-			if (!M || !M.current)
+			if(!M || !M.current)
 				score["opkilled"]++
 				continue
 			var/turf/T = M.current.loc
-			if (T && istype(T.loc, /area/security/brig)) score["arrested"] += 1
-			else if (M.current.stat == 2) score["opkilled"]++
-		if(foecount == score["arrested"]) score["allarrested"] = 1
+			if(T && istype(T.loc, /area/security/brig))
```
Comment (clusterfack) [inline code/game/gamemodes/scoreboard.dm]: There's a brig area on the escape shuttle

File: code/game/gamemodes/scoreboard.dm [hunk 9]
```diff
@@ -275,15 +334,21 @@
...
-			if (NUKE.r_code == "Nope") continue
+			if(NUKE.r_code == "Nope")
+				continue
 			var/turf/T = NUKE.loc
 			bombdat = T.loc
-			if (istype(T,/area/syndicate_station) || istype(T,/area/wizard_station) || istype(T,/area/solar/) || istype(T,/area)) nukedpenalty = 1000
-			else if (istype(T,/area/security/main) || istype(T,/area/security/brig) || istype(T,/area/security/armory) || istype(T,/area/security/checkpoint2)) nukedpenalty = 50000
-			else if (istype(T,/area/engine)) nukedpenalty = 100000
-			else nukedpenalty = 10000
+			if(istype(T,/area/syndicate_station) || istype(T,/area/wizard_station) || istype(T,/area/solar/) || istype(T,/area))
```
Comment (clusterfack) [inline code/game/gamemodes/scoreboard.dm]: Refactor of areas with a bitflag for all these special conditionals -when-

File: code/game/gamemodes/scoreboard.dm (hunk 8 above)
Comment (PJB3005) [inline code/game/gamemodes/scoreboard.dm]: Addition to this: the way it works is that any people on the RED flooring are counted as arrested.

File: code/game/gamemodes/scoreboard.dm [hunk 10]
```diff
@@ -275,15 +334,21 @@
 			break // Should only need one go-round, probably
 		var/nukedpenalty = 0
 		for(var/obj/machinery/nuclearbomb/NUKE in machines)
-			if (NUKE.r_code == "Nope") continue
+			if(NUKE.r_code == "Nope")
+				continue
 			var/turf/T = NUKE.loc
```
Comment (PJB3005) [inline code/game/gamemodes/scoreboard.dm]: `get_turf(NUKE)`

File: code/game/gamemodes/scoreboard.dm (hunk 9 above)
Comment (PJB3005) [inline code/game/gamemodes/scoreboard.dm]: WAIT A SECOND DYLAN.

You're doing ISTYPES FOR AREAS ON TURFS?

THAT'S ALWAYS FALSE.

Comment (PJB3005) [inline code/game/gamemodes/scoreboard.dm]: Also, EVERY AREA WILL RETURN 1 IN `istype(T,/area)`

File: code/game/gamemodes/scoreboard.dm (hunk 6 above)
Comment (dylanstrategie) [inline code/game/gamemodes/scoreboard.dm]: Borrowed from Cult code

File: code/game/gamemodes/scoreboard.dm (hunk 7 above)
Comment (dylanstrategie) [inline code/game/gamemodes/scoreboard.dm]: I attempted this and it did not compile, I'm not certain if this proc outputs correctly

File: code/modules/events/event_dynamic.dm [hunk 11]
```diff
@@ -99,6 +99,8 @@ var/list/event_last_fired = list()
 	//and start working via the constructor.
 	new picked_event
 
+	score["eventsendured"]++
```
Comment (clusterfack) [inline code/modules/events/event_dynamic.dm]: Even the events that don't even impact the round harmfully get points? This seems silly.

File: code/game/gamemodes/scoreboard.dm (hunk 9 above)
Comment (dylanstrategie) [inline code/game/gamemodes/scoreboard.dm]: Some day I guess

File: code/game/gamemodes/scoreboard.dm (hunk 6 above)
Comment (PJB3005) [inline code/game/gamemodes/scoreboard.dm]: Does that mean cult code is good.

Sorry but copying from caelcode isn't an excuse to write cael tier code.

File: code/modules/events/event_dynamic.dm (hunk 11 above)
Comment (dylanstrategie) [inline code/modules/events/event_dynamic.dm]: Technically speaking, a random event is a random event. And outside of the lottery I'm pretty sure they are all harmful in some fashion

File: code/game/gamemodes/scoreboard.dm (hunk 5 above)
Comment (dylanstrategie) [inline code/game/gamemodes/scoreboard.dm]: What do you mean by this ? Should I loop through a particular list ?

File: code/modules/events/event_dynamic.dm (hunk 11 above)
Comment (PJB3005) [inline code/modules/events/event_dynamic.dm]: TECHNICALLY this is true

File: code/game/gamemodes/scoreboard.dm (hunk 10 above)
Comment (dylanstrategie) [inline code/game/gamemodes/scoreboard.dm]: Fixed, but somehow Github isn't closing

File: code/game/gamemodes/scoreboard.dm (hunk 8 above)
Comment (dylanstrategie) [inline code/game/gamemodes/scoreboard.dm]: Excluded Nuke Ops from the scope of this PR, refer to OP

File: code/game/gamemodes/scoreboard.dm (hunk 6 above)
Comment (dylanstrategie) [inline code/game/gamemodes/scoreboard.dm]: It works though, and I removed a locate() this way

File: code/game/objects/effects/decals/cleanable.dm [hunk 12]
```diff
@@ -13,4 +13,29 @@
...
+	if(istype(M, /obj/effect/decal/cleanable/flour))
+		score["mess"]++
+	if(istype(M, /obj/effect/decal/cleanable/tomato_smudge))
+		score["mess"]++
+	if(istype(M, /obj/effect/decal/cleanable/egg_smudge))
+		score["mess"]++
+	if(istype(M, /obj/effect/decal/cleanable/pie_smudge))
+		score["mess"]++
+	if(istype(M, /obj/effect/decal/cleanable/soot))
+		score["mess"]++
```
Comment (PJB3005) [inline code/game/objects/effects/decals/cleanable.dm]: WHAT.

THE.

FUCK.

File: code/game/objects/explosion.dm [hunk 13]
```diff
@@ -37,7 +37,10 @@
 
 		var/start = world.timeofday
 		epicenter = get_turf(epicenter)
-		if(!epicenter) return
+		if(!epicenter)
+			return
+
+		score["explosions"]++ //For the scoreboard
```
Comment (PJB3005) [inline code/game/objects/explosion.dm]: You should probably scale this, an assistant with an IED is AS bad as a fucking 7dev

Comment (dylanstrategie) [inline code/game/objects/explosion.dm]: This is intended for the Weird category (does not affect the score). Kinda like a benchmark

File: code/modules/events/event_dynamic.dm (hunk 11 above)
Comment (clusterfack) [inline code/modules/events/event_dynamic.dm]: Yes but you put events ENDURED, hardly an accurate statement to say they 'endured' pda spam

Comment (dylanstrategie) [inline code/modules/events/event_dynamic.dm]: I could always change it to less powerful terminology like "Random Events this Shift", but it'll be fine for the most part, plus it counts positively

File: code/game/objects/explosion.dm (hunk 13 above)
Comment (PJB3005) [inline code/game/objects/explosion.dm]: ah okay then

**Comment by dylanstrategie** [comment]
Addressing all gamemode-specific scoreboard code (in which I include the nuke) wasn't planned in this PR, otherwise everything else looks to be working (maybe outside of Diseases, but it shouldn't cause problems from what testing showed)

Please go ahead with Needs Final Decisions unless there's more issues in the non-gamemode-specific code

File: code/game/gamemodes/scoreboard.dm [hunk 14]
```diff
@@ -112,114 +105,152 @@
...
-				else if (istype(T,/area/security/main) || istype(T,/area/security/brig) || istype(T,/area/security/armory) || istype(T,/area/security/checkpoint2)) nukedpenalty = 50000
-				else if (istype(T,/area/engine)) nukedpenalty = 100000
-				else nukedpenalty = 10000
-
-	if (ticker.mode.config_tag == "revolution")
+		if(score["nuked"])
+			for(var/obj/machinery/nuclearbomb/nuke in machines)
+				if(nuke.r_code == "Nope")
+					continue
+				var/turf/T = get_turf(nuke)
```
Comment (ComicIronic) [inline code/game/gamemodes/scoreboard.dm]: NOT FIXED
STILL CHECKING TURF WITH AREA TYPES

File: code/game/gamemodes/scoreboard.dm (hunk 9 above)
Comment (dylanstrategie) [inline code/game/gamemodes/scoreboard.dm]: Now commented

File: code/game/gamemodes/scoreboard.dm (hunk 10 above)
Comment (dylanstrategie) [inline code/game/gamemodes/scoreboard.dm]: Now commented

File: code/game/gamemodes/scoreboard.dm (hunk 7 above)
Comment (clusterfack) [inline code/game/gamemodes/scoreboard.dm]: Then you messed up because this is a list

**Comment by dylanstrategie** [comment]
Seeing how fucking broken that code was, this isn't a huge loss. Feel free to re-introduce those items if you feel they are interesting, but they're not worth the trouble and the massive round-end delay for the round-end scoreboard

**Pull request closed without merging**